### PR TITLE
internal: cache build info to reduce lock contention in GetGitMetadataTags

### PR DESCRIPTION
**Pull request by darccio**
### What does this PR do?

Introduces caching code to avoid calling `debug.ReadBuildInfo` multiple times each time a span is finished. It's cached on package initialisation.

Note: it's important to know that it's not possible to test `debug.ReadBuildInfo` because test binaries have no debug info, according to [golang/go#33976](https://github.com/golang/go/issues/33976).

### Motivation

This was detected as an important lock contention area using our Continuous Profiler on one of our busiest services. It should help to reduce lock contention, with potential reduction around 15%.

image

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.

Unsure? Have a question? Request a review!

### CODE REVIEW

**Comment by darccio** [comment]
Reverting to draft to investigate a bit more.

File: internal/gitmetadata.go [hunk 1]
```diff
@@ -76,11 +82,13 @@ func getTagsFromDDTags() map[string]string {
 	}
 }
 
-// getTagsFromBinary extracts git metadata from binary metadata
-func getTagsFromBinary() map[string]string {
+// getTagsFromBinary extracts git metadata from binary metadata.
+// It assumes it's running under a locked mutex.
```
[the remark this answers is not in the document]
Comment (darccio) [reply internal/gitmetadata.go:86]: @RomainMuller You are right, it was from a previous attempt that cached when the function was called.

**Review by nsrip-dd — APPROVED** [review]
Thanks!

**Review by mtoffl01 — APPROVED** [review]
nice approach!

**Pull request merged**